lib/puppet/defaults.rb

```diff
@@ -1400,6 +1400,18 @@ def self.initialize_default_settings!(settings)
       be used by _all_ environments like those in the `basemodulepath`. The only
       difference is that modules in the `basemodulepath` are pluginsynced, while
       vendored modules are not",
+      :call_hook => :on_initialize_and_write,
+      :hook => proc do |dir|
```

Comment:
I'm not sure I'm following the Puppet::Server::Settings code correctly, but it looks like these procs might get called if a thread tries to set a setting? https://github.com/puppetlabs/puppetserver/blob/ba5006b631bf4d746a99ac9e10320726836212e1/src/ruby/puppetserver-lib/puppet/server/settings.rb#L25

Reply:
Oh, that could be bad. I didn't put this in the same place as the other bit because it didn't recognize the setting when i was doing the tests there. Open to suggestions...

Reply:
I don't know, were these unit tests or in a repl? Wondering if it could be a weird interaction between how we [clear settings](https://github.com/puppetlabs/puppet/blob/904c8a52d960197173b5985d509b75237a39f348/lib/puppet/settings.rb#L1298) between tests, how we [initialize them](https://github.com/puppetlabs/puppet/blob/904c8a52d960197173b5985d509b75237a39f348/lib/puppet/test/test_helper.rb#L229) before tests, and how we [register settings when loading puppet.rb](https://github.com/puppetlabs/puppet/blob/904c8a52d960197173b5985d509b75237a39f348/lib/puppet.rb#L136).